viewing_party/party.py

```diff
@@ -1,23 +1,159 @@
+            user_data["watchlist"].remove(movie)
+            user_data["watched"].append(movie)
+            return user_data
+    return user_data    
 
 # -----------------------------------------
 # ------------- WAVE 2 --------------------
 # -----------------------------------------
+def get_watched_avg_rating(user_data):
+    # It Calculates and returns the average rating of all movies in the watched list
+    avg_rating = 0.0
+    if not user_data["watched"]:
+        return avg_rating
+    for movie in user_data["watched"]:
+        avg_rating += movie["rating"]
+    return avg_rating / len(user_data["watched"])
+
+
+def get_most_watched_genre(user_data):
+    # It returns the genre that is the most frequently watched
+    genre_list = []
+    counter = 0
+    popular_genre = None
+    if len(user_data["watched"]) == 0:
+        return None
+    else:
+        for movie in user_data["watched"]:
+            genre_list.append(movie["genre"])
 
+        for item in genre_list:
+            freq = genre_list.count(item)
+            if freq > counter:
+                counter = freq
+                popular_genre = item
+        return popular_genre
+               
 
 # -----------------------------------------
 # ------------- WAVE 3 --------------------
 # -----------------------------------------
+def get_unique_watched(user_data):
+# returns a list of dict movies the user has watched, but none of their friends have watched
+    unique_movies = []
+    all_friends_movies = [] 
+
+    for friend in user_data["friends"]:
+        for movie in friend["watched"]:
+            if movie not in all_friends_movies:
+                all_friends_movies.append(movie)
+
+    for movie in user_data["watched"]:
+        if movie not in all_friends_movies:
+            unique_movies.append(movie)
+    return unique_movies
+
+
+def get_friends_unique_watched(user_data):
+    # Returns list of dicts movies at least one of the user's friends have watched, but the user has not watched
+    user_movies = []
+    unique_friend_movies = []
+    for movie in user_data["watched"]:
+        if movie not in user_movies:
+            user_movies.append(movie)
+
+    for friend in user_data["friends"]:
+        for movie in friend["watched"]:
+            if movie not in user_movies and movie not in unique_friend_movies:
+                unique_friend_movies.append(movie)
+    return unique_friend_movies
+
+
 
-        
 # -----------------------------------------
 # ------------- WAVE 4 --------------------
 # -----------------------------------------
+def get_available_recs(user_data):
+    # Returns a list of recommended movies that the user has not watched, at least one of the user's friends has watched
+    # and the host of the movies has to be in the users subscription 
+    list_rec_movies = []
+    sub_list = []
+    unique_watched = get_friends_unique_watched(user_data)
+    
+    for sub in user_data["subscriptions"]:
+        sub_list.append(sub)

```

Comment:
Same kind of idea as earlier, this code is simply copying over a list that already exists, so it is redundant. Because we are not changing user_data["subscriptions"] at any point, we can use it as is on line 122 instead of sub_list.